### PR TITLE
Add option to populate approle credentials at specified vault path

### PR DESCRIPTION
**Pull request by dwelch0**
ticket: https://issues.redhat.com/browse/APPSRE-6389

NOTE: majority of total line changes is due to `data.json` update (testing dependency). This was required to include relevant `outputPath` attribute but also included all new changes within qontract-schemas since last generation of bundle.
* the most relevant file for review is `toplevel/role/approle-creds.go`

This PR adds ability for vault manager to read the credentials for a managed approle and output the creds to a desired path within vault.
* Additionally, refactor to reading/writing secrets was included in order to support KV engines v1 and v2. This was refactored as necessity to output approle creds to all available KVs.

Dependent on https://github.com/app-sre/qontract-schemas/pull/270

### CODE REVIEW

File: toplevel/role/approle-creds.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+			}
+			secret, err := vault.ReadSecret(address, role.OutputPath, version)
+			if err != nil {
+				log.WithFields(log.Fields{
+					"name":       role.Name,
+					"path":       role.OutputPath,
+					"kv_version": kvVersions[fmt.Sprint(pathRoot, "/")],
+					"instance":   address,
+				}).Info("[Vault Approle] Unable to read desired output path")
+				return err
```
Comment (patjlm) [inline toplevel/role/approle-creds.go:56]: what happens here if the secret does not exist? do we get a `err != nil` , in which case we should rather continue?

Comment (patjlm) [reply toplevel/role/approle-creds.go:56]: Ah, I see we'd get `nil, nil`.. sorry

File: pkg/vault/client.go [hunk 2]
```diff
@@ -438,6 +449,19 @@ func GetGroupInfo(instanceAddr string, name string) (map[string]interface{}, err
 	return entity.Data, nil
 }
 
+// write secret to vault
+func WriteApproleSecretID(instanceAddr, secretPath string) (*api.Secret, error) {
+	secret, err := getClient(instanceAddr).Logical().Write(secretPath, map[string]interface{}{})
```
Comment (patjlm) [inline pkg/vault/client.go:454]: what data do we write here? an empty map ?

Comment (patjlm) [reply pkg/vault/client.go:454]: or is this actually retrieving the role secret data in order to store it later ? It just looks weird we need to use `Write`...

Comment (dwelch0) [reply pkg/vault/client.go:454]: The `write` was a strange concept for me as well. This is based upon https://www.vaultproject.io/docs/auth/approle#configuration , specifically the `vault write -f auth/approle/role/my-role/secret-id` that is required to generate a secret id for a particular role. I will update the comments around this function to better elaborate.

Comment (dwelch0) [reply pkg/vault/client.go:454]: The [api approach](https://www.vaultproject.io/docs/auth/approle#via-the-api-1) may be more clear. They're making a `POST` to the secret-id endpoint with an empty body to generate a new `secret_id`. However, the `Write()` function of the Go vault package requires a data payload parameter, so an empty map is passed.

Comment (patjlm) [reply pkg/vault/client.go:455]: understood, thanks for the explanation. Would it be worth renaming the `WriteApproleSecretID` method to something clearer, like `GenerateApproleSecret` or something like that ?

Comment (dwelch0) [reply pkg/vault/client.go:455]: Great point. Naming it after what's technically occurring is not very helpful in this case :grimacing:

**Review by patjlm — APPROVED** [review]

**Pull request merged**